e2e/utils/content-type-builder.js

```diff
@@ -0,0 +1,316 @@
+import { expect } from '@playwright/test';
+
+export async function createComponent({ page, existingCategory, category, displayName }) {
+  await page.getByRole('button', { name: 'Create new component' }).click();
+  await page.getByLabel('Display name').fill(displayName);
+
+  // Open the category select
+  await page.getByPlaceholder('Select or enter a value').click();
+
+  if (existingCategory) {
+    await page.getByLabel(existingCategory).click();
+  } else {
+    await page.getByPlaceholder('Select or enter a value').fill(category);
+    await page.getByPlaceholder('Select or enter a value').press('Enter');
+  }
+
+  await page.getByRole('button', { name: 'Continue' }).click();
+
+  return () => deleteComponent({ page, displayName });
+}
+
+export async function deleteComponent({ page, displayName }) {
+  await page.getByRole('link', { name: displayName }).click();
+
+  // The strapi update notifier alert might be displayed in front
+  // of the edit button, which prevents the click from working
+  if (await page.getByLabel('Close').isVisible()) {
+    await page.getByLabel('Close').click();
+  }
+
+  await page.getByRole('button', { name: 'Edit' }).click();
+  await page.getByRole('button', { name: 'Delete' }).click();
+
+  await waitForReload({ page });
+}
+
+export async function waitForReload({ page }) {
+  // the restart watcher is not very reliable and often doesn't catch the server
+  // has restarted, so this trying to reload the page after 20 seconds and try
+  // again
+  try {
+    await expect(page.locator('text=Waiting for restart...')).toHaveCount(0, { timeout: 20000 });
+  } catch (error) {
+    await page.reload();
+    await expect(page.locator('text=Waiting for restart...')).toHaveCount(0, { timeout: 20000 });
+  }
+}
```

Comment:
@alexandrebodin this is the only bigger problem atm with this branch. Locally playwright goes into the catch clause and executes it, but in the CI this seems to fail straight away.

The overall problem with the reload watcher is, that it is not reliable and often the server-restart is not picked up by the frontend, which is why I need a retry. I have observed that the restart takes about ~9 seconds in the CI, so 20 seconds is generous enough.

The code to make this work must express:

- wait for the reload spinner to disappear for max. n seconds
- if it is still present: reload the page and try again

I have tried numerous approaches but couldn't make one work. Do you have a better idea?

Reply:
No better ideas. I think it's fine. What I see is the timeout being exceeded is the default 30sec one probably due to lack of resources on the CI and starvation. Just testing with the test.slow() to see if this can be mitigated 1st. But this looks alright anyway